Remove my `scalar_copy_backend_type` optimization attempt

scottmcm opened the pull request:
I added this back in https://github.com/rust-lang/rust/pull/111999 , but I no longer think it's a good idea
- It had to get scaled back to only power-of-two things to not break a bunch of targets
- LLVM seems to be getting better at memcpy removal anyway
- Introducing vector instructions has seemed to sometimes (https://github.com/rust-lang/rust/pull/115515#issuecomment-1750069529) make autovectorization worse

So this removes it from the codegen crates entirely, and instead just tries to use <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/traits/builder/trait.BuilderMethods.html#method.typed_place_copy> instead of direct `memcpy` so things will still use load/store when a type isn't `OperandValue::Ref`.

scottmcm committed:
Remove my `scalar_copy_backend_type` optimization attempt

I added this back in 111999, but I no longer think it's a good idea
- It had to get scaled back to only power-of-two things to not break a bunch of targets
- LLVM seems to be getting better at memcpy removal anyway
- Introducing vector instructions has seemed to sometimes (115515) make autovectorization worse

So this removes it from the codegen crates entirely, and instead just tries to use <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/traits/builder/trait.BuilderMethods.html#method.typed_place_copy> instead of direct `memcpy` so things will still use load/store for immediates.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b72e5ad9062f1192dfc60645050ed04836cd2cd9 with merge ab08738b5d1c784475d9fd734165d671e5617689...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b72e5ad9062f1192dfc60645050ed04836cd2cd9","merge_sha":"ab08738b5d1c784475d9fd734165d671e5617689"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8478556637/job/23233132603)
Build commit: ab08738b5d1c784475d9fd734165d671e5617689 (`ab08738b5d1c784475d9fd734165d671e5617689`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8478556637/job/23233132603"},"merge_sha":"ab08738b5d1c784475d9fd734165d671e5617689"} -->

rust-timer commented:
Queued ab08738b5d1c784475d9fd734165d671e5617689 with parent 760e567af5398a0d8c512f904e551e1f38e00d79, future [comparison URL](https://perf.rust-lang.org/compare.html?start=760e567af5398a0d8c512f904e551e1f38e00d79&end=ab08738b5d1c784475d9fd734165d671e5617689).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (ab08738b5d1c784475d9fd734165d671e5617689): [comparison URL](https://perf.rust-lang.org/compare.html?start=760e567af5398a0d8c512f904e551e1f38e00d79&end=ab08738b5d1c784475d9fd734165d671e5617689&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=760e567af5398a0d8c512f904e551e1f38e00d79&end=ab08738b5d1c784475d9fd734165d671e5617689&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.1%  | [1.1%, 1.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.7%, -0.6%] | 4     |
| Improvements ✅ <br /> (secondary) | -1.4% | [-1.4%, -1.4%] | 1     |
| All ❌✅ (primary)                 | -0.6% | [-0.7%, -0.6%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=760e567af5398a0d8c512f904e551e1f38e00d79&end=ab08738b5d1c784475d9fd734165d671e5617689&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=760e567af5398a0d8c512f904e551e1f38e00d79&end=ab08738b5d1c784475d9fd734165d671e5617689&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.3% | [-2.7%, -1.4%] | 4     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=760e567af5398a0d8c512f904e551e1f38e00d79&end=ab08738b5d1c784475d9fd734165d671e5617689&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.1%, -0.0%] | 13    |
| Improvements ✅ <br /> (secondary) | -0.9% | [-1.3%, -0.0%] | 36    |
| All ❌✅ (primary)                 | -0.1% | [-0.1%, -0.0%] | 13    |
</details>

**Bootstrap**: 667.865s -> 668.349s (0.07%)
**Artifact size**: 315.66 MiB -> 315.57 MiB (-0.03%)

compiler-errors commented on compiler/rustc_codegen_ssa/src/mir/operand.rs:
```diff
@@ -419,7 +418,14 @@ impl<'a, 'tcx, V: CodegenObject> OperandValue<V> {
                     return;
                 }
-                base::memcpy_ty(bx, dest.llval, dest.align, r, source_align, dest.layout, flags)
+                bx.memcpy_known_size(
```
🤔 why isn't this also just using `typed_place_copy`?

also, `memcpy_known_size` has one callsite (this one) -- you should inline it. i don't really see the value of exposing it if it's only being used once, seems to add more confusion to the api imo.

fee1-dead commented:
r? compiler

compiler-errors commented:
I already started this review anyways

r? compiler-errors

compiler-errors commented on compiler/rustc_codegen_ssa/src/traits/builder.rs:
```diff
@@ -280,17 +280,31 @@ pub trait BuilderMethods<'a, 'tcx>:
             let temp = self.load_operand(src);
-            temp.val.store(self, dst);
+            temp.val.store_with_flags(self, dst, flags);
+        } else if flags.contains(MemFlags::NONTEMPORAL) {
+            // HACK(nox): This is inefficient but there is no nontemporal memcpy.
+            let ty = self.backend_type(dst.layout);
+            let val = self.load(ty, src.llval, src.align);
+            self.store_with_flags(val, dst.llval, dst.align, flags);
```
What is the difference between this and the above code? It feels subtly different but only barely

compiler-errors replied:
nontemporal sounds scary enough that i feel like we should reorder these branches, then? like, check nontemporal first? unless you're particularly convinced there will be no side effects.

scottmcm committed:
Put the `NONTEMPORAL` case first

That's how it was in `store_with_flags` before this PR, so let's do that here too just to be sure we get the right thing.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8549579688/job/23425151183) [(plain)](https://github.com/rust-lang/rust/commit/dc006701c7b9a3872a8951b13e441278d4e89551/checks/23425151183/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#17 exporting to docker image format
#17 sending tarball 29.1s done
#17 DONE 34.5s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
```

</details>

compiler-errors commented:
@bors r+

bors commented:
:pushpin: Commit 2a41a32d21912800cbb62047a057a39dbf80d5ee has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 2a41a32d21912800cbb62047a057a39dbf80d5ee -->
<!-- homu: {"type":"Approved","sha":"2a41a32d21912800cbb62047a057a39dbf80d5ee","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 2a41a32d21912800cbb62047a057a39dbf80d5ee with merge 456e99a399f4f448a7da21aea9520410479d9969...
<!-- homu: {"type":"BuildStarted","head_sha":"2a41a32d21912800cbb62047a057a39dbf80d5ee","merge_sha":"456e99a399f4f448a7da21aea9520410479d9969"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8611781005/job/23599628294) [(plain)](https://github.com/rust-lang-ci/rust/commit/456e99a399f4f448a7da21aea9520410479d9969/checks/23599628294/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/issues/issue-122805.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-122805/issue-122805.ll" "/checkout/tests/codegen/issues/issue-122805.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/checkout/tests/codegen/issues/issue-122805.rs:12:11: error: CHECK: expected string not found in input
/checkout/tests/codegen/issues/issue-122805.rs:12:11: error: CHECK: expected string not found in input
// CHECK: insertelement <8 x i16>
          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-122805/issue-122805.ll:7:32: note: scanning from here
define dso_local void @convert(ptr noalias nocapture noundef writeonly sret([16 x i8]) align 1 dereferenceable(16) %_0, ptr noalias nocapture noundef readonly align 2 dereferenceable(16) %value) unnamed_addr #0 {
                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-122805/issue-122805.ll:11:5: note: possible intended match here
 %1 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 1

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-122805/issue-122805.ll
Check file: /checkout/tests/codegen/issues/issue-122805.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'issue_122805.15b653d76730f15b-cgu.0' 
            2: source_filename = "issue_122805.15b653d76730f15b-cgu.0" 
            3: target datalayout = "e-m:e-p:32:32-p10:8:8-p20:8:8-i64:64-n32:64-S128-ni:1:10:20" 
            4: target triple = "wasm32-unknown-wasi" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite) 
            7: define dso_local void @convert(ptr noalias nocapture noundef writeonly sret([16 x i8]) align 1 dereferenceable(16) %_0, ptr noalias nocapture noundef readonly align 2 dereferenceable(16) %value) unnamed_addr #0 { 
check:12'0                                    X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:12'0     ~~~~~~~
check:12'0     ~~~~~~~
            9:  %self = load i16, ptr %value, align 2, !noundef !1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  %0 = tail call i16 @llvm.bswap.i16(i16 %self) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           11:  %1 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:12'1         ?                                                             possible intended match
           12:  %self1 = load i16, ptr %1, align 2, !noundef !1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           13:  %2 = tail call i16 @llvm.bswap.i16(i16 %self1) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14:  %3 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 2 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           15:  %self2 = load i16, ptr %3, align 2, !noundef !1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16:  %4 = tail call i16 @llvm.bswap.i16(i16 %self2) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           17:  %5 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 3 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           18:  %self3 = load i16, ptr %5, align 2, !noundef !1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           19:  %6 = tail call i16 @llvm.bswap.i16(i16 %self3) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           20:  %7 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 4 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21:  %self4 = load i16, ptr %7, align 2, !noundef !1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           22:  %8 = tail call i16 @llvm.bswap.i16(i16 %self4) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           23:  %9 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 5 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24:  %self5 = load i16, ptr %9, align 2, !noundef !1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           25:  %10 = tail call i16 @llvm.bswap.i16(i16 %self5) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           26:  %11 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 6 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           27:  %self6 = load i16, ptr %11, align 2, !noundef !1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           28:  %12 = tail call i16 @llvm.bswap.i16(i16 %self6) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           29:  %13 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 7 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           30:  %self7 = load i16, ptr %13, align 2, !noundef !1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           31:  %14 = tail call i16 @llvm.bswap.i16(i16 %self7) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           32:  store i16 %0, ptr %_0, align 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           33:  %addr16.sroa.2.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 2 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           34:  store i16 %2, ptr %addr16.sroa.2.0._0.sroa_idx, align 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           35:  %addr16.sroa.3.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 4 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           36:  store i16 %4, ptr %addr16.sroa.3.0._0.sroa_idx, align 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           37:  %addr16.sroa.4.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 6 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           38:  store i16 %6, ptr %addr16.sroa.4.0._0.sroa_idx, align 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           39:  %addr16.sroa.5.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 8 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           40:  store i16 %8, ptr %addr16.sroa.5.0._0.sroa_idx, align 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           41:  %addr16.sroa.6.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 10 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           42:  store i16 %10, ptr %addr16.sroa.6.0._0.sroa_idx, align 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           43:  %addr16.sroa.7.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 12 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           44:  store i16 %12, ptr %addr16.sroa.7.0._0.sroa_idx, align 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45:  %addr16.sroa.8.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 14 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46:  store i16 %14, ptr %addr16.sroa.8.0._0.sroa_idx, align 1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           47:  ret void 
check:12'0     ~~~~~~~~~~
           48: } 
check:12'0     ~~
check:12'0     ~
check:12'0     ~
           50: ; Function Attrs: mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none) 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51: declare hidden i16 @llvm.bswap.i16(i16) #1 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:12'0     ~
check:12'0     ~
           53: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite) "target-cpu"="generic" } 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54: attributes #1 = { mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none) } 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:12'0     ~
check:12'0     ~
           56: !llvm.ident = !{!0} 
check:12'0     ~~~~~~~~~~~~~~~~~~~~
check:12'0     ~
check:12'0     ~
           58: !0 = !{!"rustc version 1.79.0-nightly (456e99a39 2024-04-09)"} 
check:12'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59: !1 = !{} 
check:12'0     ~~~~~~~~~
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8611781005/job/23600488573)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8611781005/job/23600488573","builder_name":"checks-actions"} -->

compiler-errors commented:
@rustbot author

scottmcm committed:
Update 122805 test for PR 123185

bors commented:
:pushpin: Commit 556b47ee1ab265442094658738cd5e1c31c97c69 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 556b47ee1ab265442094658738cd5e1c31c97c69 -->
<!-- homu: {"type":"Approved","sha":"556b47ee1ab265442094658738cd5e1c31c97c69","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

DianQK commented on tests/codegen/issues/issue-122805.rs:
```diff
@@ -9,15 +14,27 @@
+// OPT3WINX64: load <8 x i16>
+// OPT3WINX64-NEXT: call <8 x i16> @llvm.bswap
+// OPT3WINX64-NEXT: store <8 x i16>
 // CHECK-NEXT: ret void
```
Ah, thanks! I hadn't noticed that different optimization levels yield different optimization effects.
I believe it makes sense to differentiate between O2 and O3 here.

The changes to O2 seem fine. I'll check again on this later.

DianQK replied:
I confirm that is fine. I'm just curious why the results are the same for O2 and O3 is in LLVM's opt. 
I've found that Rust's O2 disables LLVM's SLP vectorization pass:  https://github.com/rust-lang/rust/blob/f96442b4482fef6a50b65f2284bcfb8921320183/compiler/rustc_codegen_ssa/src/back/write.rs#L248-L249.

dianqk reviewed (commented):
> LLVM seems to be getting better at memcpy removal anyway.

I think we can expect to see more significant changes after https://github.com/llvm/llvm-project/pull/87190.

bors commented:
:hourglass: Testing commit 556b47ee1ab265442094658738cd5e1c31c97c69 with merge ba787eef3c0cf4f0b1c35d51e468dfb6040651f8...
<!-- homu: {"type":"BuildStarted","head_sha":"556b47ee1ab265442094658738cd5e1c31c97c69","merge_sha":"ba787eef3c0cf4f0b1c35d51e468dfb6040651f8"} -->

rust-log-analyzer commented:
The job **`dist-i586-gnu-i586-i686-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/8627162963/job/23646598465) [(plain)](https://github.com/rust-lang-ci/rust/commit/ba787eef3c0cf4f0b1c35d51e468dfb6040651f8/checks/23646598465/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [codegen] tests/codegen/issues/issue-119422.rs ... ignored, only executed when the pointer width is 64bit ((because the LLVM type of i64 for usize shows up))
test [codegen] tests/codegen/issues/issue-105386-ub-in-debuginfo.rs ... ok
test [codegen] tests/codegen/issues/issue-103327.rs ... ok
test [codegen] tests/codegen/issues/issue-103840.rs ... ok
test [codegen] tests/codegen/issues/issue-122805.rs#OPT3WINX64 ... ignored, only executed when the operating system is windows (only-x86_64)
test [codegen] tests/codegen/issues/issue-115385-llvm-jump-threading.rs ... ok
test [codegen] tests/codegen/issues/issue-116878.rs ... ok
test [codegen] tests/codegen/issues/issue-121719-common-field-offset.rs ... ok
test [codegen] tests/codegen/issues/issue-122805.rs#OPT3LINX64 ... FAILED
---
test [codegen] tests/codegen/sanitizer/address-sanitizer-globals-tracking.rs#ASAN-FAT-LTO ... ok

failures:

---- [codegen] tests/codegen/issues/issue-122805.rs#OPT3LINX64 stdout ----

error in revision `OPT3LINX64`: verification with 'FileCheck' failed
status: exit status: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-122805.OPT3LINX64/issue-122805.ll" "/checkout/tests/codegen/issues/issue-122805.rs" "--check-prefix=CHECK" "--check-prefix" "OPT3LINX64" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/checkout/tests/codegen/issues/issue-122805.rs:32:16: error: OPT3LINX64: expected string not found in input
// OPT3LINX64: load <8 x i16>
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-122805.OPT3LINX64/issue-122805.ll:7:22: note: scanning from here
define void @convert(ptr noalias nocapture noundef writeonly sret([16 x i8]) align 1 dereferenceable(16) %_0, ptr noalias nocapture noundef readonly align 2 dereferenceable(16) %value) unnamed_addr #0 {
                     ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-122805.OPT3LINX64/issue-122805.ll:11:24: note: possible intended match here
 %1 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 1


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/issues/issue-122805.OPT3LINX64/issue-122805.ll


-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
            1: ; ModuleID = 'issue_122805.15b653d76730f15b-cgu.0' 
            2: source_filename = "issue_122805.15b653d76730f15b-cgu.0" 
            3: target datalayout = "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-i128:128-f64:32:64-f80:32-n8:16:32-S128" 
            4: target triple = "i586-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite) uwtable 
            7: define void @convert(ptr noalias nocapture noundef writeonly sret([16 x i8]) align 1 dereferenceable(16) %_0, ptr noalias nocapture noundef readonly align 2 dereferenceable(16) %value) unnamed_addr #0 { 
check:32'0                          X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
check:32'0     ~~~~~~~
check:32'0     ~~~~~~~
            9:  %self = load i16, ptr %value, align 2, !noundef !2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           10:  %0 = tail call i16 @llvm.bswap.i16(i16 %self) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           11:  %1 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:32'1                            ?                                          possible intended match
           12:  %self1 = load i16, ptr %1, align 2, !noundef !2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           13:  %2 = tail call i16 @llvm.bswap.i16(i16 %self1) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           14:  %3 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           15:  %self2 = load i16, ptr %3, align 2, !noundef !2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           16:  %4 = tail call i16 @llvm.bswap.i16(i16 %self2) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           17:  %5 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 3 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           18:  %self3 = load i16, ptr %5, align 2, !noundef !2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           19:  %6 = tail call i16 @llvm.bswap.i16(i16 %self3) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           20:  %7 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 4 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           21:  %self4 = load i16, ptr %7, align 2, !noundef !2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           22:  %8 = tail call i16 @llvm.bswap.i16(i16 %self4) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           23:  %9 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 5 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           24:  %self5 = load i16, ptr %9, align 2, !noundef !2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           25:  %10 = tail call i16 @llvm.bswap.i16(i16 %self5) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           26:  %11 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 6 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           27:  %self6 = load i16, ptr %11, align 2, !noundef !2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           28:  %12 = tail call i16 @llvm.bswap.i16(i16 %self6) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           29:  %13 = getelementptr inbounds [8 x i16], ptr %value, i32 0, i32 7 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           30:  %self7 = load i16, ptr %13, align 2, !noundef !2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           31:  %14 = tail call i16 @llvm.bswap.i16(i16 %self7) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           32:  store i16 %0, ptr %_0, align 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           33:  %addr16.sroa.2.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 2 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           34:  store i16 %2, ptr %addr16.sroa.2.0._0.sroa_idx, align 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           35:  %addr16.sroa.3.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 4 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           36:  store i16 %4, ptr %addr16.sroa.3.0._0.sroa_idx, align 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           37:  %addr16.sroa.4.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 6 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           38:  store i16 %6, ptr %addr16.sroa.4.0._0.sroa_idx, align 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           39:  %addr16.sroa.5.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 8 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           40:  store i16 %8, ptr %addr16.sroa.5.0._0.sroa_idx, align 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           41:  %addr16.sroa.6.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 10 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           42:  store i16 %10, ptr %addr16.sroa.6.0._0.sroa_idx, align 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           43:  %addr16.sroa.7.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 12 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           44:  store i16 %12, ptr %addr16.sroa.7.0._0.sroa_idx, align 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45:  %addr16.sroa.8.0._0.sroa_idx = getelementptr inbounds i8, ptr %_0, i32 14 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46:  store i16 %14, ptr %addr16.sroa.8.0._0.sroa_idx, align 1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           47:  ret void 
check:32'0     ~~~~~~~~~~
           48: } 
check:32'0     ~~
check:32'0     ~
check:32'0     ~
           50: ; Function Attrs: mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none) 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51: declare i16 @llvm.bswap.i16(i16) #1 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:32'0     ~
check:32'0     ~
           53: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(argmem: readwrite) uwtable "probe-stack"="inline-asm" "target-cpu"="pentium" } 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           54: attributes #1 = { mustprogress nocallback nofree nosync nounwind speculatable willreturn memory(none) } 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:32'0     ~
check:32'0     ~
           56: !llvm.module.flags = !{!0} 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
           57: !llvm.ident = !{!1} 
check:32'0     ~~~~~~~~~~~~~~~~~~~~
check:32'0     ~
check:32'0     ~
           59: !0 = !{i32 8, !"PIC Level", i32 2} 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60: !1 = !{!"rustc version 1.79.0-nightly (ba787eef3 2024-04-10)"} 
check:32'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           61: !2 = !{} 
check:32'0     ~~~~~~~~~
------------------------------------------



```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8627162963/job/23647491959)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/8627162963/job/23647491959","builder_name":"checks-actions"} -->

bors commented:
:pushpin: Commit 593e900ad25f1b21cc218ea8bcce3c5e3e94ceac has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors 593e900ad25f1b21cc218ea8bcce3c5e3e94ceac -->
<!-- homu: {"type":"Approved","sha":"593e900ad25f1b21cc218ea8bcce3c5e3e94ceac","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 593e900ad25f1b21cc218ea8bcce3c5e3e94ceac with merge c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c...
<!-- homu: {"type":"BuildStarted","head_sha":"593e900ad25f1b21cc218ea8bcce3c5e3e94ceac","merge_sha":"c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/8634706949/job/23675368807)
Approved by: compiler-errors
Pushing c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"compiler-errors","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/8634706949/job/23675368807"},"merge_sha":"c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c"} -->

rust-timer commented:
Finished benchmarking commit (c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c): [comparison URL](https://perf.rust-lang.org/compare.html?start=5974fe87c4d711949caa64fc1e8366685c8fc190&end=c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=5974fe87c4d711949caa64fc1e8366685c8fc190&end=c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.7% | [-0.7%, -0.7%] | 4     |
| Improvements ✅ <br /> (secondary) | -2.0% | [-2.7%, -1.2%] | 2     |
| All ❌✅ (primary)                 | -0.7% | [-0.7%, -0.7%] | 4     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=5974fe87c4d711949caa64fc1e8366685c8fc190&end=c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=5974fe87c4d711949caa64fc1e8366685c8fc190&end=c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.5% | [-1.5%, -1.5%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=5974fe87c4d711949caa64fc1e8366685c8fc190&end=c2239bca5b89a8d3573cc0fc0f2fa65c50edb79c&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.0%] | 14    |
| Improvements ✅ <br /> (secondary) | -0.9% | [-1.4%, -0.0%] | 36    |
| All ❌✅ (primary)                 | -0.1% | [-0.2%, -0.0%] | 14    |
</details>

**Bootstrap**: 675.526s -> 675.28s (-0.04%)
**Artifact size**: 318.49 MiB -> 318.45 MiB (-0.01%)
